content/blog/2022-05-06-end-to-end-computer-vision-api-part-3-remote-exp-ci-cd.md

```diff
@@ -0,0 +1,292 @@
+
+      ```yaml
+      ---
+      train-model:
+        needs: deploy-runner
+        runs-on: [self-hosted, cml-runner]
+        container:
+          image: iterativeai/cml:0-dvc2-base1
+          options: --gpus all
+        environment: cloud
+        permissions:
+          contents: read
+          id-token: write
+        steps:
+          - uses: actions/checkout@v2
+          - uses: actions/setup-python@v2
+            with:
+              python-version: '3.9'
+      ```
+
+   3. rerun the entire DVC pipeline & push metrics back to GitHub;
+
+      ```yaml
+      ...
+      - name: dvc-repro-cml
+              env:
+                REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+              run: |
+                # Install dependencies
+                pipenv install --skip-lock
+                pipenv run dvc pull
+                pipenv run dvc exp run
+                pipenv run dvc push
+      ...
+      ```
+
+   4. open a pull request and post a report to it that contains a table with
+      metrics and model outputs on a few test images:
+
+      ```bash
+      ...
+      # Open a pull request
+      cml pr dvc.lock metrics.json training_metrics.json training_metrics_dvc_plots/**
+      # Create CML report
+      echo "## Metrics" >> report.md
+      pipenv run dvc metrics show --md >> report.md
+      echo "## A few random test images" >> report.md
+      for file in $(ls data/test_preds/ |sort -R |tail -20); do
+        cml publish data/test_preds/$file --md >> report.md
+      done
+      cml send-comment --pr --update report.md
+      ...
+      ```
+
+      The report structure is fully customizable. Below is an example of what
+      the PR and the CML report would look like in this case. The test images
+      show (from left to right) input images, ground truth masks and prediction
+      masks.
+
+      ![PR and CML report](/uploads/images/2022-05-06/pr_cml_report.png '=800')
+
+   At this point, we can assess the results in Iterative Studio and GitHub and
+   decide whether we want to accept the PR or keep experimenting.
+
+2. [Workflow for deploying to the development environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/2-develop.yaml)
+
+   ![Workflow for deploying to the development environment](/uploads/images/2022-05-06/workflow_dev.png '=800')
+
+   Once we are happy with our model's performance on the experiment branch, we
+   can merge it into the development branch. This would trigger a different
+   CI/CD job that will:
+
+   1. retrain the model if the `dev` branch contains changes not present in the
+      `exp` branch. DVC will skip this stage if that's not the case. This step
+      looks almost identical to the third step in the above workflow.
+   2. deploy the web REST API application (that incorporates the new model) to a
+      development endpoint on Heroku:
+
+      ```yaml
+      ---
```

Comment:
```suggestion
      ...
```

Reply:
Tried it, but `restyled-io` wants to change it back

Reply:
tbh I think maybe it's better to remove all the `...` in all the snippets? Seems unneeded. Basically:

```
... # unneeded
code:
  - some
  ...  # needed, something missing in-between
  - text
... # unneeded
```